File: main-service/todo/todo.py
import json
from flask import Blueprint, g, jsonify, request

from todo.client import TodoClient
from middleware.middleware import token_required

todo_blueprint = Blueprint('todo_blueprint', __name__)


@todo_blueprint.route('/create', methods=['POST'])
@token_required
def create_todo():
    try:
        user_data = g.user_data
        title = request.json.get('title')
        description = request.json.get('description')
        print(user_data, title, description)
        client = TodoClient()
        result = client.create_todo(title=title,
                                    description=description,
                                    user_id=user_data.get('ID'))
        return jsonify({'success': True, 'message': result.message})
    except:
        return jsonify({'success': False, 'message': result.message})


@todo_blueprint.route('/get', methods=['GET'])
@token_required
def get_todos():
    try:
        user_data = g.user_data
        client = TodoClient()
        result = client.get_todos(user_id=user_data.get('ID'))
        return jsonify({
            'success': True,
            'message': result.message,
            'data': json.loads(result.data)
        })
    except:
        return jsonify({
            'success': False,
            'message': result.message,
            'data': []
        })


@todo_blueprint.route('/update', methods=['PUT'])
@token_required
def update_todo():
    try:
        id = request.json.get('id')
        title = request.json.get('title')
        description = request.json.get('description')
        client = TodoClient()
        result = client.update_todo(title=title,
                                    description=description,
                                    id=id)
        return jsonify({'success': True, 'message': result.message})
    except:
        return jsonify({
            'success': False,
            'message': result.message,
            'data': []
        })


@todo_blueprint.route('/delete/<int:id>', methods=['DELETE'])
@token_required
def delete_todo(id):
    try:
        client = TodoClient()
        result = client.delete_todo(id)
        return jsonify({
            'success': False,
            'message': result.message,
        })
    except:
        return jsonify({
            'success': False,
            'message': result.message,
        })

File: web/src/api/todo.js
import { TodoServiceClient } from './todo_grpc_web_pb'
import { CreateTodoRequest, GetTodoRequest, GetTodosRequest, UpdateTodoRequest, DeleteTodoRequest } from './todo_pb'

const client = new TodoServiceClient('http://localhost:2000')

const getTodos = (userId, handleGetTodos) => {
  const request = new GetTodosRequest()
  request.setUserid(userId)
  client.getTodos(request, null, (err, response) => {
    if (err) {
      console.log(err)
    } else {
      console.log(response)
      const result = response.getData()
      handleGetTodos(JSON.parse(result))
    }
  })
}

const getTodo = (userId, todoId) => {
  const request = new GetTodoRequest()
  request.setUserid(userId)
  request.setItemid(todoId)
  client.getTodo(request, null, (err, response) => {
    if (err) {
      console.log(err)
    } else {
      console.log(response)
    }
  })
}

const createTodo = (userId, title, description) => {
  const request = new CreateTodoRequest()
  request.setUserid(userId)
  request.setTitle(title)
  request.setDescription(description)
  client.createTodo(request, null, (err, response) => {
    if (err) {
      console.log("create todo err", err)
    } else {
      console.log("create todo response", response)
    }
  })
}

const updateTodo = (todoId, title, description) => {
  const request = new UpdateTodoRequest()
  request.setId(todoId)
  request.setTitle(title)
  request.setDescription(description)
  client.updateTodo(request, null, (err, response) => {
    if (err) {
      console.log(err)
    } else {
      console.log(response)
    }
  })
}

const deleteTodo = (todoId) => {
  const request = new DeleteTodoRequest()
  request.setId(todoId)
  client.deleteTodo(request, null, (err, response) => {
    if (err) {
      console.log(err)
    } else {
      console.log(response)
    }
  })
}


export {
  getTodos,
  getTodo,
  createTodo,
  updateTodo,
  deleteTodo,
}